microprofile/server/src/main/java/io/helidon/microprofile/server/ServerCdiExtension.java

```diff
@@ -582,4 +591,89 @@ private HttpRouting.Builder findRouting(String className,
 
         return serverNamedRoutingBuilder(routingName);
     }
+
+    // -- CDI bean producers for ServerRequest and ServerResponse -----------------------
+
+    private void afterBeanDiscovery(@Observes AfterBeanDiscovery event) {
+        event.addBean(new ServerRequestProducer());
```

Comment:
You absolutely can do it this way. You can also do it fluently:
```
event.addBean()
  .qualifiers(Set.of(Default.Literal.INSTANCE, Any.Literal.INSTANCE)) // technically not needed; always supplied implicitly if you do nothing else
  .addTransitiveTypeClosure(ServerRequest.class)
  .scope(RequestScoped.class)
  .createWith(cc -> Contexts.context().orElseThrow().get(ServerRequest.class).orElseThrow());
```
…if I've typed it right; see [`BeanConfigurator`](https://jakarta.ee/specifications/cdi/3.0/apidocs/jakarta/enterprise/inject/spi/configurator/beanconfigurator) for details.

Reply:
That is a lot shorter, I think.